ltm/agent.py

```diff
@@ -96,6 +125,121 @@ def sanitize_string(s: str) -> str:
+        log_llm_call(self.run_name, self.save_name, self.debug_level, label=f"should_update-{self.llm_call_idx}")
+        colour_print("YELLOW", f"Should update: {response_text}")
+
+        if "Should Update: No." in response_text:
+            return
+        elif "Should Update: Yes." not in response_text:
+            return
+
+        # Produce a new one
+        context.append(make_assistant_message(response_text))
+        context.append(make_user_message(new_task_def))
+
+        try:
+            response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_callback, temperature=self.temperature)
+            log_llm_call(self.run_name, self.save_name, self.debug_level, label=f"do_update-{self.llm_call_idx}")
+
+            new_task = sanitize_and_parse_json(response)
+            self.process_new_task(new_task)
+
+            print(f"Added Task:\n{json.dumps(new_task, indent=2)}\n")
+
+        except Exception as e:
+            print("Encountered exception when creating task: " + repr(e))
+
+    def process_new_task(self, new_task: dict):
+        new_task["trigger_details"]["triggered"] = False
+        if new_task["trigger_details"]["trigger_condition"] == "MESSAGE_COUNT":
+            new_task["trigger_details"]["trigger_value"] = int(new_task["trigger_details"]["trigger_value"])
+
+        if new_task["trigger_details"]["trigger_condition"] == "MESSAGE_CONTENT":
+            new_task["trigger_details"]["trigger_value"] = new_task["trigger_details"]["trigger_value"].lower()
+
+        if new_task["trigger_details"]["trigger_condition"] == "TIME":
+            new_task["trigger_details"]["trigger_value"] = time.mktime(datetime.datetime.strptime(new_task["trigger_details"]["trigger_value"], "%Y-%m-%d_%H:%M").timetuple())
+
+        self.task_memory.append(new_task)
+
+    def remove_completed_tasks(self):
+
+        for task in self.task_memory:
+            if task["trigger_details"]["triggered"]:
+                self.task_memory.remove(task)
+
+    def update_task(self, task, context):
+
+        cond = task["trigger_details"]["trigger_condition"]
+        val = task["trigger_details"]["trigger_value"]
+        if cond == "MESSAGE_COUNT":
+            # Decrement the message counter and ready it to be triggered
+            task["trigger_details"]["trigger_value"] -= 1
+            task["trigger_details"]["triggered"] = task["trigger_details"]["trigger_value"] <= 0
+
+        elif cond == "MESSAGE_CONTENT":
+            user_message = context[-1]["content"].lower()
+            task["trigger_details"]["triggered"] = val in user_message
+
+        elif cond == "TIME":
+            task["trigger_details"]["triggered"] = self.now.timestamp() > val
+
+        if task["trigger_details"]["triggered"]:
+            print(f"{cond}: Task triggered: {json.dumps(task, indent=2)}")
+
+        return task["trigger_details"]["triggered"]
```

Comment:
This looks better and it's easier to read than `process_new_task`. You could maybe do `details = task["trigger_details"]` too.